### PR TITLE
[JENKINS-40693, JENKINS-43713] upgrade to jetty 9.4.x

### PR DESCRIPTION
**Pull request by olamy**
Signed-off-by: olivier lamy <olamy@apache.org>

see https://issues.jenkins-ci.org/browse/JENKINS-40693

### CODE REVIEW

**Comment by aheritier** [comment]
+1

**Comment by daniel-beck** [comment]
Is there a complete changelog for these versions in Jetty?

**Comment by oleg-nenashev** [comment]
@daniel-beck @aheritier @olamy  https://github.com/eclipse/jetty.project/compare/jetty-9.2.15.v20160210...jetty-9.2.21.v20170120 There are many changes I'd say.

But these changes include `Jenkinsfile`, so what could possibly go wrong? :)

**Comment by oleg-nenashev** [comment]
Summon @jenkinsci/code-reviewers since we need some extra eyes on it

**Comment by paladox** [comment]
Why not go with 9.3.17.v20170317 since jenkins uses java 8 now. Also the change log is here for 9.2 https://github.com/eclipse/jetty.project/blob/jetty-9.2.x/VERSION.txt

and for 9.3 https://github.com/eclipse/jetty.project/blob/jetty-9.3.x/VERSION.txt

**Comment by oleg-nenashev** [comment]
@paladox Yes, I was going to create a PR for Jetty 9.3 on the next week. But this change may be still useful before we evaluate Jetty 9.3

**Comment by paladox** [comment]
Ok. Thanks.

**Comment by oleg-nenashev** [comment]
Created https://issues.jenkins-ci.org/browse/JENKINS-43501 for the upgrade to 9.3

**Comment by oleg-nenashev** [comment]
@olamy Do you have any insights about possible compat issues between 9.2 and 9.3?

**Comment by olamy** [comment]
@oleg-nenashev I don't think so (but as you know we never know :-) )
worth the upgrade.
Let me know I can change my pr.

**Comment by oleg-nenashev** [comment]
@olamy Please do. The main effort is to test the things (Jenkins core manual tests + ATH), I commit to do it within 1 week

**Comment by olamy** [comment]
well 9.4.x will be a better option! I will test that.

**Comment by olamy** [comment]
made some progress. (weird unit test fail on my osx but not on a linux box...).
I introduced a new class (maybe this one will be added to Jetty source code).
I will let you know.
Maybe I could rename the branch :-)

**Comment by olamy** [comment]
https://github.com/eclipse/jetty.project/issues/1481

**Comment by olamy** [comment]
for the record, to fix unit test: ``` ifconfig lo0 alias 127.0.0.2 ```

**Comment by olamy** [comment]
so ATM this depends on jetty 9.4.5-SNAPSHOT version (but at least can be used for testing purpose)

File: src/java/winstone/Launcher.java [hunk 1]
```diff
@@ -352,6 +352,7 @@ public static void main(String argv[]) throws IOException {
             new Launcher(args);
         } catch (Throwable err) {
             Logger.log(Logger.ERROR, RESOURCES, "Launcher.ContainerStartupError", err);
+            System.exit(1);
```
Comment (oleg-nenashev) [inline src/java/winstone/Launcher.java]: Unrelated, but OK. Maybe deserves a separate changelog entry if there is a real issue with it

**Comment by olamy** [comment]
@oleg-nenashev have setup ` ifconfig lo0 alias 127.0.0.2 `
For some reason (I don't know) there are unit tests using this ip.....
I didn't want to change that.

File: pom.xml [hunk 2]
```diff
@@ -2,7 +2,7 @@
   <modelVersion>4.0.0</modelVersion>
   <groupId>org.jenkins-ci</groupId>
   <artifactId>winstone</artifactId>
-  <version>3.4-SNAPSHOT</version>
+  <version>3.5-SNAPSHOT</version>
```
Comment (olamy) [inline pom.xml]: LGTM I will do

**Comment by olamy** [comment]
@oleg-nenashev I hope new Jetty release around end of next week.

File: README.md [hunk 3]
```diff
@@ -166,3 +166,8 @@ New features in v2.0:
 * The engine is now Jetty 8.x, instead of the original from-scratch implementation
 
 For pre-1.0 history, see [the upstream changelog](http://winstone.sourceforge.net/#recent)
+
+## Development
+If you have some unit test failures you may add an interface/ip alias such
+
+``` sudo ifconfig lo0 alias 127.0.0.2 ```
```
Comment (oleg-nenashev) [inline README.md]: OK, fine for me

File: pom.xml [hunk 4]
```diff
@@ -301,17 +308,19 @@
       <version>${jetty.version}</version>
     </dependency>
     -->
-
+    <!--
+    spdy abandonned
```
Comment (oleg-nenashev) [inline pom.xml]: Any potential impact on transient deps?

File: pom.xml [hunk 5]
```diff
@@ -228,6 +224,17 @@
       <id>repo.jenkins-ci.org</id>
       <url>http://repo.jenkins-ci.org/public/</url>
     </repository>
+    <!-- can be removed when jetty 9.4.5 released -->
```
Comment (oleg-nenashev) [inline pom.xml]: Should be removed now I suppose

File: src/testwebapp/WEB-INF/web.xml [hunk 6]
```diff
@@ -1,9 +1,10 @@
 <?xml version="1.0" encoding="ISO-8859-1"?>
-<!DOCTYPE web-app
-    PUBLIC "-//Sun Microsystems, Inc.//DTD Web Application 2.3//EN"
-    "http://java.sun.com/dtd/web-app_2_3.dtd">
+<web-app
+		xmlns="http://java.sun.com/xml/ns/javaee"
+		xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+		xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_2_5.xsd"
```
Comment (oleg-nenashev) [inline src/testwebapp/WEB-INF/web.xml]: Would be nice to have schema validation in the PR builder. Oh wait, PR builder first...

File: pom.xml (hunk 5 above)
Comment (oleg-nenashev) [inline pom.xml]: remove?

File: pom.xml (hunk 4 above)
Comment (oleg-nenashev) [inline pom.xml]: is there an impact on transient deps?

File: src/testwebapp/WEB-INF/web.xml (hunk 6 above)
Comment (olamy) [inline src/testwebapp/WEB-INF/web.xml]: ? I don't get it? well it's a test resource

File: pom.xml (hunk 4 above)
Comment (olamy) [inline pom.xml]: Nope I don't think so.
FYI I have some test locally to enable http2 as well. But not fully test so will do an other pr later.

**Comment by olamy** [comment]
ready to merge I guess?
I'm happy to push the button :-)
Let me know. I think I have to fix something for maven-hudson-dev-plugin :-)

**Comment by daniel-beck** [comment]
Could we perhaps use this opportunity to update https://github.com/jenkinsci/winstone/blob/master/src/java/winstone/LocalStrings.properties#L1 ? It should have been a filtered resource long ago…

**Comment by oleg-nenashev** [comment]
@jglick Could you please cut the release or approve https://github.com/jenkins-infra/repository-permissions-updater/pull/293/files ?

**Pull request merged**